hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -8,9 +8,449 @@ import is.hail.expr.types.physical._
+  def joinPoint[T: ParameterPack](implicit ctx: EmitStreamContext): DefinableJoinPoint[T] = ctx.jb.joinPoint[T](ctx.mb)
+
+  private case class Source[+A](setup0: Code[Unit], close0: Code[Unit], setup: Code[Unit], close: Code[Unit], firstPull: Option[Code[Ctrl]], pull: Code[Ctrl])
+
+  abstract class Stream[+A] {
+    private[CodeStream] def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A]
+
+    def fold[S: ParameterPack](s0: S, f: (A, S) => S, ret: S => Code[Ctrl])(implicit ctx: EmitStreamContext): Code[Ctrl] =
+      CodeStream.fold(this, s0, f, ret)
+    def forEach(f: A => Code[Unit], ret: Code[Ctrl])(implicit ctx: EmitStreamContext): Code[Ctrl] =
+      CodeStream.forEach(this, f, ret)
+    def forEach(mb: MethodBuilder)(f: A => Code[Unit]): Code[Unit] =
+      CallCC[Unit]((jb, ret) => CodeStream.forEach(this, f, ret(()))(EmitStreamContext(mb, jb)))
+    def mapCPS[B](
+      f: (EmitStreamContext, A, B => Code[Ctrl]) => Code[Ctrl],
+      setup0: Option[Code[Unit]] = None,
+      setup:  Option[Code[Unit]] = None,
+      close0: Option[Code[Unit]] = None,
+      close:  Option[Code[Unit]] = None
+    ): Stream[B] = CodeStream.mapCPS(this)(f, setup0, setup, close0, close)
+    def map[B](
+      f: A => B,
+      setup0: Option[Code[Unit]] = None,
+      setup:  Option[Code[Unit]] = None,
+      close0: Option[Code[Unit]] = None,
+      close:  Option[Code[Unit]] = None
+    ): Stream[B] = CodeStream.map(this)(f, setup0, setup, close0, close)
+    def flatMap[B](f: A => Stream[B]): Stream[B] =
+      CodeStream.flatMap(map(f))
+  }
+
+  implicit class StreamPP[A](val stream: Stream[A]) extends AnyVal {
+    def filter(cond: A => Code[Boolean])(implicit pp: ParameterPack[A]): Stream[A] =
+      CodeStream.filter(stream, cond)
+  }
+
+  def unfold[A, S: ParameterPack](
+    s0: S,
+    f: (S, EmitStreamContext, COption[(A, S)] => Code[Ctrl]) => Code[Ctrl]
+  ): Stream[A] = new Stream[A] {
+   def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+      val s = newLocal[S]
+      Source[A](
+        setup0 = s.init,
+        close0 = Code._empty,
+        setup = s := s0,
+        close = Code._empty,
+        firstPull = None,
+        pull = f(s.load, ctx, _.apply(
+          none = eos,
+          // Warning: `a` should not depend on `s`
+          some = { case (a, s1) => Code(s := s1, push(a)) })))
+    }
+  }
+
+  def range(start: Code[Int], step: Code[Int], len: Code[Int]): Stream[Code[Int]] =
+    unfold[Code[Int], (Code[Int], Code[Int])](
+      s0 = (start, len),
+      f = { case ((cur, rem), _ctx, k) =>
+        implicit val ctx = _ctx
+        val xCur = newLocal[Code[Int]]
+        val xRem = newLocal[Code[Int]]
+        Code(
+          xCur := cur,
+          xRem := rem - 1,
+          k(COption(xRem.load < 0,
+                    (xCur.load, (xCur.load + 1, xRem.load)))))
```

Comment:
This is wrong, it won't handle overflow correctly.  We have battle-tested range code, you should reuse it.

Reply:
Okay, I will.

Reply:
Which range code are you referring to?

The emit rule for range will need to assert that the length is not too long, and that (start + step * len) is within Int range. Given that precheck, I'm not seeing where the overflow can happen.

(That should also be xCur + step; I'll fix that.)

Reply:
> The emit rule for range will need to assert that the length is not too long

You can't do that.  Range it well-defined for all values of start, stop and step.  It looks like the current code is broken.  I thought the last test in testArrayRange would catch this.  Can you add a test that will catch this?

You can find a (hopefully) correct implementation in an older version in the Emit code for ArrayRange.  e047f540ce5a4e16dd41a6ee2d9f1237371f76cb has it, for example.

Reply:
Are you referring to https://github.com/hail-is/hail/blob/e047f540ce5a4e16dd41a6ee2d9f1237371f76cb/hail/src/main/scala/is/hail/expr/ir/Emit.scala#L1966
?

That looks the same as the current `StreamRange` emit code. In particular, it errors on `ArrayRange(start, stop, 1)` if `stop - start > MaxInt`.

I think I see a way to make it work for all arguments: keep only a single state, which is the next output. That makes the termination check more complicated: if `step > 0`, end if `cur > stop`, else end if `cur < stop`. Are you thinking something like this?

Maybe there's a way to avoid that branch. One idea is to arrange it so that step is always positive, by multiplying all outputs by +/- 1 (so replace `StreamRange(start, stop, -d)` by `StreamRange(-stop, -start, d) * -1`), though that still doesn't work if `stop = MinInt`.

Reply:
Actually, maybe using the two-part state (next value, and number of remaining elements) can work if we treat the latter as an unsigned int. Then `rem > 0` doesn't work, but `rem != 0` does.

Reply:
Ah, my mistake, this is fine.  You called this range but take a start, step, len, while range everywhere else takes a start, end, step.

I think the length check is necessary until we start support 64-bit arrays.